src/main/scala/controllers/ActionHandler.scala

```diff
@@ -83,65 +82,86 @@ class ActionHandler @Inject()(applications: ApplicationOps, applicationForms: Ap
     }
   }
 
-  def RedirectToPreview(id: ApplicationId, sectionNumber: Int): Future[Result] = {
-    Future.successful(Redirect(controllers.routes.ApplicationPreviewController.previewSection(id, sectionNumber)))
-  }
+  def CompleteAndPreview(id: ApplicationId, sectionNumber: Int, fieldValues: JsObject): Future[Result] = {
 
-  def renderSectionForm(id: ApplicationId,
-                        sectionNumber: Int,
-                        section: Option[ApplicationSection],
-                        questions: Map[String, Question],
-                        fields: Seq[Field],
-                        errs: FieldErrors,
-                        hints: FieldHints): Future[Result] = {
-    val answers = section.map { s => JsonHelpers.flatten("", s.answers) }.getOrElse(Map[String, String]())
-
-    gatherApplicationDetails(id).map {
-      case Some((app, appForm, opp)) => selectSectionForm(sectionNumber, section, questions, answers, fields, errs, app, appForm, opp)
-      case None => NotFound
+    sectionTypeFor(sectionNumber) match {
+      case VanillaSection =>
+        val errs = check(fieldValues, previewChecksFor(sectionNumber))
+        if (errs.isEmpty) {
+          JsonHelpers.allFieldsEmpty(fieldValues) match {
+            case true => applications.deleteSection(id, sectionNumber).map { _ =>
+              Redirect(routes.ApplicationController.show(id))
+            }
+            case false => applications.completeSection(id, sectionNumber, fieldValues).flatMap {
+              case Nil => Future.successful( Redirect(routes.ApplicationPreviewController.previewSection(id, sectionNumber)))
+              case errs => redisplaySectionForm(id, sectionNumber, JsonHelpers.flatten("", fieldValues), errs)
+            }
+          }
+        }
+        else redisplaySectionForm(id, sectionNumber, JsonHelpers.flatten("", fieldValues), errs)
+      case CostSection => Future.successful(wip(sectionFormCall(id, sectionNumber).url))
     }
   }
 
-  def redisplaySectionForm(id: ApplicationId, sectionNumber: Int, answers: Map[String, String], errs: FieldErrors = noErrors): Future[Result] = {
```

Comment:
Can you just check your IDEA settings - look at Editor->Code Style->Scala and check "Tabs and Indents". Should be don't use tabs and the indents should be 2